From 1499031ff3ed3bafb725d47dfb49aea39ceb912b Mon Sep 17 00:00:00 2001
From: Zheng Yejian <zhengyejian1@huawei.com>
Date: Thu, 8 Sep 2022 19:52:58 +0000
Subject: [PATCH 0236/1076] [Huawei] livepatch: Check calltrace of idle tasks

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187517

--------------------------------

for_each_process_thread() only include init_task and its descendant,
but not include idle tasks. Idle tasks also require calltraces checking.

Signed-off-by: Zheng Yejian <zhengyejian1@huawei.com>
---
 arch/arm/kernel/livepatch.c        | 6 ++++++
 arch/arm64/kernel/livepatch.c      | 6 ++++++
 arch/powerpc/kernel/livepatch_32.c | 6 ++++++
 arch/powerpc/kernel/livepatch_64.c | 6 ++++++
 arch/x86/kernel/livepatch.c        | 7 ++++++-
 5 files changed, 30 insertions(+), 1 deletion(-)

diff --git a/arch/arm/kernel/livepatch.c b/arch/arm/kernel/livepatch.c
index 9afef2f32edc..38958d378a81 100644
--- a/arch/arm/kernel/livepatch.c
+++ b/arch/arm/kernel/livepatch.c
@@ -228,6 +228,7 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 {
 	int ret;
 	struct task_struct *g, *t;
+	unsigned int cpu;
 
 	for_each_process_thread(g, t) {
 		if (klp_is_migration_thread(t->comm))
@@ -236,6 +237,11 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 		if (ret)
 			return ret;
 	}
+	for_each_online_cpu(cpu) {
+		ret = check_task_calltrace(idle_task(cpu), args, fn);
+		if (ret)
+			return ret;
+	}
 	return 0;
 }
 
diff --git a/arch/arm64/kernel/livepatch.c b/arch/arm64/kernel/livepatch.c
index e23e6597fc1e..d7ce303a1f6c 100644
--- a/arch/arm64/kernel/livepatch.c
+++ b/arch/arm64/kernel/livepatch.c
@@ -224,6 +224,7 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 {
 	int ret;
 	struct task_struct *g, *t;
+	unsigned int cpu;
 
 	for_each_process_thread(g, t) {
 		if (klp_is_migration_thread(t->comm))
@@ -232,6 +233,11 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 		if (ret)
 			return ret;
 	}
+	for_each_online_cpu(cpu) {
+		ret = check_task_calltrace(idle_task(cpu), args, fn);
+		if (ret)
+			return ret;
+	}
 	return 0;
 }
 
diff --git a/arch/powerpc/kernel/livepatch_32.c b/arch/powerpc/kernel/livepatch_32.c
index b6bba306d525..a6f0d6cdea79 100644
--- a/arch/powerpc/kernel/livepatch_32.c
+++ b/arch/powerpc/kernel/livepatch_32.c
@@ -255,6 +255,7 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 {
 	int ret;
 	struct task_struct *g, *t;
+	unsigned int cpu;
 
 	for_each_process_thread(g, t) {
 		if (klp_is_migration_thread(t->comm))
@@ -263,6 +264,11 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 		if (ret)
 			return ret;
 	}
+	for_each_online_cpu(cpu) {
+		ret = check_task_calltrace(idle_task(cpu), args, fn);
+		if (ret)
+			return ret;
+	}
 	return 0;
 }
 
diff --git a/arch/powerpc/kernel/livepatch_64.c b/arch/powerpc/kernel/livepatch_64.c
index 2e9a89a0edfc..b3f0cf505356 100644
--- a/arch/powerpc/kernel/livepatch_64.c
+++ b/arch/powerpc/kernel/livepatch_64.c
@@ -276,6 +276,7 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 {
 	int ret;
 	struct task_struct *g, *t;
+	unsigned int cpu;
 
 	for_each_process_thread(g, t) {
 		if (klp_is_migration_thread(t->comm))
@@ -284,6 +285,11 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 		if (ret)
 			return ret;
 	}
+	for_each_online_cpu(cpu) {
+		ret = check_task_calltrace(idle_task(cpu), args, fn);
+		if (ret)
+			return ret;
+	}
 	return 0;
 }
 
diff --git a/arch/x86/kernel/livepatch.c b/arch/x86/kernel/livepatch.c
index 4c8c1f0802ea..c8da05117f6e 100644
--- a/arch/x86/kernel/livepatch.c
+++ b/arch/x86/kernel/livepatch.c
@@ -282,6 +282,7 @@ static int do_check_calltrace(bool (*fn)(void *, int *, unsigned long), void *da
 {
 	int ret = 0;
 	struct task_struct *g, *t;
+	unsigned int cpu;
 
 	for_each_process_thread(g, t) {
 		if (klp_is_migration_thread(t->comm))
@@ -291,7 +292,11 @@ static int do_check_calltrace(bool (*fn)(void *, int *, unsigned long), void *da
 		if (ret)
 			return ret;
 	}
-
+	for_each_online_cpu(cpu) {
+		ret = check_task_calltrace(idle_task(cpu), fn, data);
+		if (ret)
+			return ret;
+	}
 	return 0;
 }
 
-- 
2.17.1

